### PR TITLE
tests/periph_rtt_min: add test to figure out RTT_MIN_OFFSET

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds a simple test to figure out the correct value for `RTT_MIN_VALUE`

IMO this is also a good reason to make `RTT_MIN_VALUE` public and not only in `ztimer` code.

### Testing procedure

`BOARD=<board>make -C tests/periph_rtt_min/ flash test`

```
main(): This is RIOT! (Version: 2020.07-devel-774-g243984-pr_rtt_min_value_test)
Evaluate RTT_MIN_VALUE over 1024 samples
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
RTT_MIN_VALUE for samr21-xpro: 10
```

```
main(): This is RIOT! (Version: 2020.07-devel-774-g243984-pr_rtt_min_value_test)
Evaluate RTT_MIN_VALUE over 1024 samples
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
RTT_MIN_VALUE for iotlab-m3: 2

```

You can also see that if in `tests/periph_rtt` `TICKS_TO_WAIT` is set lower than `RTT_MIN_VALUE` the test does not work.

Note hardware specification may suggest a value different than the one resolved by the test, e.g. `cc2538` points out a value of 5. It will also depend on the `RTT_FREQUENCY`, should be ran at the highest possible frequency.

### Issues/PRs references

Figure out while debugging why changing to `ztimer` broke support for `samr21-xpro` in https://github.com/RIOT-OS/RIOT/pull/13824.

### CODE REVIEW

File: tests/periph_rtt_min/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+## About
+
+Describe here what the test application is about
```
Comment (aabadie) [inline tests/periph_rtt_min/README.md:3]: Maybe "Describe here what the test application is about" ? :)
(and thanks for using `make generate-test` btw ;))

File: tests/periph_rtt_min/main.c [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#include <stdio.h>
+#include <stdatomic.h>
+
+#include "cpu.h"
+#include "periph_conf.h"
+#include "periph/rtt.h"
+#include "timex.h"
+#include "ztimer.h"
+
+#define MIN_SLEEP       (100)
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:30]:
```suggestion
#define MIN_SLEEP_US    (100)
```

**Comment by fjmolinas** [comment]
Rebased otherwise there would be conflict due to the `stm32` changes.

File: tests/periph_rtt_min/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+## About
+
+This applications is meant to determine RTT_MIN_VALUE for a specific BOARD.
```
Comment (benpicco) [inline tests/periph_rtt_min/README.md:3]: What is `RTT_MIN_VALUE` used for? What's it's meaning?

File: tests/periph_rtt_min/main.c [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#include <stdio.h>
+#include <stdatomic.h>
+
+#include "cpu.h"
+#include "periph_conf.h"
+#include "periph/rtt.h"
+
+#include "xtimer.h"
+
+#define MAX_WAIT_US     (100)
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:31]: Why is this needed?

File: tests/periph_rtt_min/README.md (hunk 3 above)
Comment (fjmolinas) [reply tests/periph_rtt_min/README.md:3]: I want to change it to `RTT_MIN_OFFSET` https://github.com/RIOT-OS/RIOT/pull/14259, currently this value is only used in `ztimer` to avoid setting callbacks that wont be triggered.

File: tests/periph_rtt_min/main.c (hunk 4 above)
Comment (fjmolinas) [reply tests/periph_rtt_min/main.c:31]: I wanted to wait for a minimum amount of time to account for the time between the callback being set and handled.

**Comment by benpicco** [comment]
Please squash!

**Comment by aabadie** [comment]
Please rebase and adapt the name to RTT_MIN_OFFSET instead of RTT_MIN_VALUE.

**Comment by fjmolinas** [comment]
Rebased.

**Comment by fjmolinas** [comment]
Should I change the name to `tests/periph_rtt_min_offset`?

File: tests/periph_rtt_min/main.c [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    rtt_init();
+
+    printf("Evaluate RTT_MIN_OFFSET over %" PRIu32 " samples\n",
+           (uint32_t)SAMPLES);
+
+    for (unsigned i = 0; i < SAMPLES; i++) {
+        uint32_t offset = 0;
+        cb_triggered = false;
+        while (cb_triggered == false) {
```
Comment (kaspar030) [inline tests/periph_rtt_min/main.c:59]: I'm trying to get this logic. basically it is setting an alarm, waiting the time it should take using xtimer, then verifying if the alarm triggered. IMO that's a bit brittle.
For synchronized RTTs, that'll probably work, but if the RTT works like a regular timer and just quantizes on the RTT clock, every set is late by `0<x<1` ticks, as the set_alarm() can happen anywhere within one RTT clock cycle.
This is supposed to figure out the smallest value that can be set without underflowing, right?
Would it maybe make sense to use the same logic as done for periph_timer_short_relative_set?
There, values are tried from larger to 0, until the timer clearly doesn't trigger anymore.

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c:59]: In the case of `RTT` trying from larger values down might mean waiting for a long time, some of our RTT are running at `1Hz`. And why would going down from larger be better than going up? Seems just as brittle to me.

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c:59]: ping @kaspar030!

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c:59]: Ping @kaspar030?

File: tests/periph_rtt_min/README.md (hunk 3 above)
Comment (benpicco) [reply tests/periph_rtt_min/README.md]: Better add a comment.

File: tests/periph_rtt_min/main.c (hunk 5 above)
Comment (benpicco) [reply tests/periph_rtt_min/main.c:59]: How about using `xtimer_mutex_lock_timeout()` instead?

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c]: Did that, also found a bug in stm32f1 rtt code.

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c]: Will split the stm32 change out.

**Comment by fjmolinas** [comment]
Split the stm32 fix out

File: tests/periph_rtt_min/main.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#ifndef SAMPLES
+#define SAMPLES 1024LU
+#endif
+
+static mutex_t _mtx = MUTEX_INIT_LOCKED;
+
+void cb(void *arg)
+{
+    (void)arg;
+    mutex_unlock(&_mtx);
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:44]:
```suggestion
    mutex_unlock(arg);
```

File: tests/periph_rtt_min/main.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+/* min. amount of time to wait between set_alarm() */
+#define MIN_WAIT_US     (100UL)
+#define US_PER_TICK     (US_PER_SEC / RTT_FREQUENCY)
+
+#ifndef SAMPLES
+#define SAMPLES 1024LU
+#endif
+
+static mutex_t _mtx = MUTEX_INIT_LOCKED;
+
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:40]:
```suggestion
```

File: tests/periph_rtt_min/main.c [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+void cb(void *arg)
+{
+    (void)arg;
+    mutex_unlock(&_mtx);
+}
+
+int main(void)
+{
+    uint32_t value = 0;
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:49]:
```suggestion
    uint32_t value = 0;
    mutex_t lock = MUTEX_INIT_LOCKED;
```

File: tests/periph_rtt_min/main.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+           (uint32_t)SAMPLES);
+
+    for (unsigned i = 0; i < SAMPLES; i++) {
+        uint32_t offset = 0;
+        int ret = -1;
+        while (ret != 0) {
+            offset++;
+            uint32_t now = rtt_get_counter();
+            mutex_trylock(&_mtx);
+            rtt_set_alarm((now + offset) % RTT_MAX_VALUE, cb, 0);
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:63]:
```suggestion
            rtt_set_alarm((now + offset) % RTT_MAX_VALUE, cb, &lock);
```

File: tests/periph_rtt_min/main.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    printf("Evaluate RTT_MIN_OFFSET over %" PRIu32 " samples\n",
+           (uint32_t)SAMPLES);
+
+    for (unsigned i = 0; i < SAMPLES; i++) {
+        uint32_t offset = 0;
+        int ret = -1;
+        while (ret != 0) {
+            offset++;
+            uint32_t now = rtt_get_counter();
+            mutex_trylock(&_mtx);
```
Comment (benpicco) [inline tests/periph_rtt_min/main.c:62]: What is that for?

Comment (fjmolinas) [reply tests/periph_rtt_min/main.c:62]: Since there are multiple iterations you always want the mutex to be locked before `xtimer_mutex_lock_timeout`, if it timed out then it would not be locked on each new "iteration". I added a comment.

Comment (benpicco) [reply tests/periph_rtt_min/main.c]: > if it timed out then it would not be locked on each new "iteration"

If `xtimer_mutex_lock_timeout()` times out the mutex is still locked.

File: tests/periph_rtt_min/main.c (hunk 5 above)
Comment (fjmolinas) [reply tests/periph_rtt_min/main.c]: > I'm trying to get this logic. basically it is setting an alarm, waiting the time it should take using xtimer, then verifying if the alarm triggered. IMO that's a bit brittle.

It is waiting always longer then the time it should take to be set, I'm always waiting at least 100ul which is about 3 ticks for a 32Khz timer, maybe I can set that too ticks related.

**Comment by fjmolinas** [comment]
@benpicco I addressed your comments.

File: tests/periph_rtt_min/main.c (hunk 10 above)
Comment (fjmolinas) [reply tests/periph_rtt_min/main.c]: True I'll remove.

File: tests/periph_rtt_min/main.c (hunk 5 above)
Comment (fjmolinas) [reply tests/periph_rtt_min/main.c]: I added a 3ticks margin to the time waited.

**Comment by benpicco** [comment]
Please squash!

File: tests/periph_rtt_min/main.c (hunk 5 above)
Comment (kaspar030) [reply tests/periph_rtt_min/main.c]: discussed offline, no issue here from my side anymore.

**Comment by fjmolinas** [comment]
@benpicco I added bbee446c08333e2e4420284690ae5ad82fd0ac18 to add some more details on the application results as well as wrap the `rtt` calls in a `irq_enable` `irq_disable` block

**Comment by benpicco** [comment]
With disabling IRQs we won't be able to spot bugs like #14779
The RTT API should be safe to use without disabling IRQs :wink:

**Comment by fjmolinas** [comment]
> With disabling IRQs we won't be able to spot bugs like #14779
> The RTT API should be safe to use without disabling IRQs wink

Removed and squashed.

**Comment by fjmolinas** [comment]
All green here @benpicco @kaspar030

**Review by benpicco — APPROVED** [review]
I just gave it another try, works fine and the code looks good.
